api/v1alpha1/validator/hostdevicenetwork_webhook.go

```diff
@@ -25,50 +27,66 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	ctrl "sigs.k8s.io/controller-runtime"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
-	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/Mellanox/network-operator/api/v1alpha1"
 )
 
 // log is for logging in this package.
 var hostDeviceNetworkLog = logf.Log.WithName("hostdevicenetwork-resource")
 
-func (w *HostDeviceNetwork) SetupWebhookWithManager(mgr ctrl.Manager) error {
+type hostDeviceNetworkValidator struct {
+	v1alpha1.HostDeviceNetwork
+}
+
+func SetupHostDeviceNetworkWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(w).
+		For(&v1alpha1.HostDeviceNetwork{}).
+		WithValidator(&hostDeviceNetworkValidator{}).
 		Complete()
 }
 
 //nolint:lll
 //+kubebuilder:webhook:path=/validate-mellanox-com-v1alpha1-hostdevicenetwork,mutating=false,failurePolicy=fail,sideEffects=None,groups=mellanox.com,resources=hostdevicenetworks,verbs=create;update,versions=v1alpha1,name=vhostdevicenetwork.kb.io,admissionReviewVersions=v1
 
-var _ webhook.Validator = &HostDeviceNetwork{}
-
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
-func (w *HostDeviceNetwork) ValidateCreate() (admission.Warnings, error) {
+func (w *hostDeviceNetworkValidator) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
 	if skipValidations {
 		nicClusterPolicyLog.Info("skipping CR validation")
 		return nil, nil
 	}
 
+	hostDeviceNetwork, ok := obj.(*v1alpha1.HostDeviceNetwork)
+	if !ok {
+		return nil, errors.New("failed to unmarshal HostDeviceNetwork object to validate")
+	}
+	w.HostDeviceNetwork = *hostDeviceNetwork
```

Comment:
can we have issues with concurrency ?

that is, webhook gets called concurently with two different requests

Reply:
I think theoretically we might, just as we probably might have had issues with the previous implementation (the nicclusterpolicy object might change during validation). Let's assess the risks and address them in a separate PR.

WDYT?

Reply:
ack, please open a task about it to check.

Reply:
@adrianchiris https://github.com/Mellanox/network-operator/issues/714